File: 카톡봇/maplegg.js
Jsoup = org.jsoup.Jsoup;
FS = FileStream;
var loc="sdcard/katalkbot/Bots/maplelog.json";

if (FS.read(loc)==null) FS.write(loc, "{}");


Number.prototype.comma = function() {
   var coma=this.toString().replace(/(?=(\d{3})+(?!\d))/g, ",");
   if(coma[0]==",") coma=coma.substr(1);
return coma;
}

function recordnick(sender,nick){
	var rd = JSON.parse(FS.read(loc)) || {};
  rd[sender] = rd[sender] || {};
  rd[sender][nick] = (rd[sender][nick] || 0) + 1;
	let n=""
	var temparr = [];
		for (i in rd[sender]){
			temparr.push(`${i}/${rd[sender][i]}`);
		}
		temparr.sort((a, b) => b.split("/")[1] - a.split("/")[1]);
		n=temparr[0].split("/")[0];
	if(rd[sender][nick]>30&&nick==n){
		rd[sender] = {};
		rd[sender][nick] = 2;
	}
	FS.write(loc, JSON.stringify(rd))
	
}

function recommendnick(sender,replier){
	var n="";
	var rd = JSON.parse(FS.read(loc));
	if(rd[sender]==undefined){
		rd[sender] = {};
		n="";
	}
	else{
		result = [];
		for (i in rd[sender]){
			result.push(i+"/"+rd[sender][i]);
		}
		result.sort((a, b)=>a.split("/")[1] - b.split("/")[1]).reverse();
		n=result[0].split("/")[0];
	}
	
	return n;
}

function responseFix(room, msg, sender, isGroupChat, replier, imageDB, packageName) {
  if(room=="바다 월드") return;
  if(msg.split(" ")[0]=="@메이플"||msg.split(" ")[0]=="!메이플") {
	try{ 
		  var nick = msg.split(" ")[1];
		  if(nick==undefined)
			  nick=recommendnick(sender,replier);
		  if(nick=="") replier.reply("닉네임을 입력해 주세요.");
		  else
		  {
			  recordnick(sender,nick);
			  
			  var today=new Date();
				var d = new Date(today.setDate(today.getDate() - 1));
				var d2=(d.getYear()+1900)+"-"+String(d.getMonth()+1).padStart(2, "0")+"-"+String(d.getDate()).padStart(2, "0");
			  
			  
			  var ocid1 = Jsoup.connect("https://open.api.nexon.com/maplestory/v1/id?character_name="+encodeURIComponent(nick)).header("Content-Type","application/json").header("x-nxopen-api-key","live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59").ignoreContentType(true).ignoreHttpErrors(true).get().text();
				var ocid2=JSON.parse(ocid1)["ocid"];

				var s1 = Jsoup.connect("https://open.api.nexon.com/maplestory/v1/character/basic?ocid="+ocid2).header("Content-Type","application/json").header("x-nxopen-api-key","live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59").ignoreContentType(true).ignoreHttpErrors(true).get().text();
				var s2=JSON.parse(s1);
				var server=s2["world_name"];


				if(server.indexOf("리부트")==(-1)){
					var answer = Jsoup.connect("https://open.api.nexon.com/maplestory/v1/ranking/overall?ocid="+ocid2+"&date="+d2).header("Content-Type","application/json").header("x-nxopen-api-key","live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59").ignoreContentType(true).ignoreHttpErrors(true).get().text();
				}
				else{
					var answer = Jsoup.connect("https://open.api.nexon.com/maplestory/v1/ranking/overall?ocid="+ocid2+"&date="+d2+"&world_type=1").header("Content-Type","application/json").header("x-nxopen-api-key","live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59").ignoreContentType(true).ignoreHttpErrors(true).get().text();
				}
					
				var t=JSON.parse(answer);
				
			  var job1 =t["ranking"][0]["class_name"];
			  var job2 =t["ranking"][0]["sub_class_name"];
			  var lv =Number(t["ranking"][0]["character_level"]);
			  var per=s2["character_exp_rate"];
			  var img=s2["character_image"];
			  var pop =t["ranking"][0]["character_popularity"];
			  var gu =t["ranking"][0]["character_guildname"];
			  if(gu==null||gu==undefined||gu=="")
				  gu="(없음)";
			  
			  if(job2!=null&&job2!=undefined&&job2!="")
				  var printjob=job1+"/"+job2;
			  else
				  var printjob=job1+"/"+job1;
			  
			  if(server.indexOf("리부트")==(-1)){
					
					var answer2= Jsoup.connect("https://open.api.nexon.com/maplestory/v1/ranking/overall?ocid="+ocid2+"&date="+d2+"&class="+job1+"-"+job2).header("Content-Type","application/json").header("x-nxopen-api-key","live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59").ignoreContentType(true).ignoreHttpErrors(true).get().text();
				}
				else{
					var answer2= Jsoup.connect("https://open.api.nexon.com/maplestory/v1/ranking/overall?ocid="+ocid2+"&date="+d2+"&class="+job1+"-"+job2+"&world_type=1").header("Content-Type","application/json").header("x-nxopen-api-key","live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59").ignoreContentType(true).ignoreHttpErrors(true).get().text();
				}
				
				var t2=JSON.parse(answer2);
			  
			  var ranking_all=Number(t["ranking"][0]["ranking"]);
			  var ranking_class=Number(t2["ranking"][0]["ranking"]);
			  
				replier.reply([
				"["+nick+"@"+server+"]",
				printjob,
				"레벨: " + lv +"("+per+"%)",
				"길드: "+ gu+" │ "+"인기도: "+ pop,
				"월드 랭킹: "+ ranking_all.comma()+"위",
				"직업 랭킹(월드): "+ ranking_class.comma()+"위",
				"\u200b".repeat(500),
				"메이플지지 링크: https://maple.gg/u/"+encodeURIComponent(nick),
				"",
				"환산 링크: https://maplescouter.com/info?name="+encodeURIComponent(nick),
				"",
				"캐릭터 이미지 다운로드",
				img,
				"",
				"Special Thanks 정쿠"
				].join("\n"));
			  
			  
		}
	}
	catch(e)
	{
		var nick = msg.substr(5);
		replier.reply("["+nick+"]\n2023.12.21 이후 기록이 없는 캐릭터명 입니다.");
	}
}



}

function onNotificationPosted(sbn, sm) {
    var packageName = sbn.getPackageName();
    if (!packageName.startsWith("com.kakao.tal")) return;
    var actions = sbn.getNotification().actions;
    if (actions == null) return;
    var userId = sbn.getUser().hashCode();
    for (var n = 0; n < actions.length; n++) {
        var action = actions[n];
        if (action.getRemoteInputs() == null) continue;
        var bundle = sbn.getNotification().extras;

        var msg = bundle.get("android.text").toString();
        var sender = bundle.getString("android.title");
        var room = bundle.getString("android.subText");
        if (room == null) room = bundle.getString("android.summaryText");
        var isGroupChat = room != null;
        if (room == null) room = sender;
        var replier = new com.xfl.msgbot.script.api.legacy.SessionCacheReplier(packageName, action, room, false, "");
        var icon = bundle.getParcelableArray("android.messages")[0].get("sender_person").getIcon().getBitmap();
        var image = bundle.getBundle("android.wearable.EXTENSIONS");
        if (image != null) image = image.getParcelable("background");
        var imageDB = new com.xfl.msgbot.script.api.legacy.ImageDB(icon, image);
        com.xfl.msgbot.application.service.NotificationListener.Companion.setSession(packageName, room, action);
        if (this.hasOwnProperty("responseFix")) {
            responseFix(room, msg, sender, isGroupChat, replier, imageDB, packageName, userId != 0);
        }
    }
}

File: iris_python/msgbot/Bots/maplegg.py
import requests
import json
from PIL import Image, ImageDraw, ImageFont
import io
import os
from urllib import parse
from bs4 import BeautifulSoup
from msgbot.Bots.maple_nickskip.nickskip_module import recordnick, recommendnick, comma, get_yesterday_date, history_db_save

#api ocid 검색
def search_api_ocid(nick):
    headers = {
        "Content-Type": "application/json",
        "x-nxopen-api-key": "live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59"
    }
    res = requests.get("https://open.api.nexon.com/maplestory/v1/id", params={"character_name": nick}, headers=headers)
    res.raise_for_status()     # 200이 아니면 에러
    #print(res.text)
    ocid1 = json.loads(res.text)
    ocid2 = ocid1["ocid"]
    return ocid2

#api 검색_basic
def search_maple_api(ocid):
    headers = {
        "Content-Type": "application/json",
        "x-nxopen-api-key": "live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59"
    }
    res = requests.get("https://open.api.nexon.com/maplestory/v1/character/basic", params={"ocid": ocid}, headers=headers)
    res.raise_for_status()     # 200이 아니면 에러
    #print(res.text)
    s2 = json.loads(res.text)
    return s2

#api 검색_ranking
def search_maple_api_ranking(ocid, isreboot):

    d2 = get_yesterday_date()

    headers = {
        "Content-Type": "application/json",
        "x-nxopen-api-key": "live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59"
    }
    
    params = {"ocid": ocid, "date": d2}
    if isreboot == 1:
        params = {"ocid": ocid, "date": d2, "world_type":1}

    res = requests.get("https://open.api.nexon.com/maplestory/v1/ranking/overall", params=params, headers=headers)
    res.raise_for_status()     # 200이 아니면 에러
    #print(res.text)
    s2 = json.loads(res.text)
    return s2

#api 검색_직업랭킹
def search_maple_api_ranking_class(ocid, job1, job2, isreboot):

    d2 = get_yesterday_date()

    headers = {
        "Content-Type": "application/json",
        "x-nxopen-api-key": "live_5fbf44d53f909c000739c6ded2630548b1340053ca172cbe0d59ef023ae9477f69320869e3c5ee348598e4c96c389f59"
    }
    
    url="https://open.api.nexon.com/maplestory/v1/ranking/overall"
    if isreboot == 1:
        url = url + f"?ocid={ocid}&date={d2}&class={job1}-{job2}&world_type=1"
    else:
        url = url + f"?ocid={ocid}&date={d2}&class={job1}-{job2}"

    res = requests.get(url, headers=headers)
    res.raise_for_status()     # 200이 아니면 에러
    #print(res.text)
    s2 = json.loads(res.text)
    return s2

def generate_character_image(image_url, nick, server, job, lv, per, guild, pop, world_rank, class_rank):

    response = requests.get(image_url)
    response.raise_for_status()

    # 1. 원본 이미지 로드
    src_img = Image.open(io.BytesIO(response.content)).convert("RGBA")
    w, h = src_img.size

    # 2. 중앙 1/4 영역 crop
    crop_w, crop_h = w // 4, h // 3
    left = (w - crop_w) // 2
    top = (h - crop_h) // 2
    right = left + crop_w
    bottom = top + crop_h

    cropped_img = src_img.crop((left, top, right, bottom))

    # 3. 새 캔버스 생성 (500x500)
    canvas = Image.new("RGBA", (300, 320), (255, 255, 255, 255))

    # 4. crop 이미지를 캔버스 왼쪽에 붙이기 (세로 중앙)
    paste_x = (300 - crop_w) // 2
    canvas.paste(cropped_img, (paste_x, 10), cropped_img)

    # 5. 텍스트 그리기
    draw = ImageDraw.Draw(canvas)

    try:
        nick_font = ImageFont.truetype("res/fonts/MaplestoryFont_OTF/Maplestory OTF Bold.otf", 25)
    except:
        nick_font = ImageFont.load_default()

    try:
        content_font = ImageFont.truetype("res/fonts/MaplestoryFont_OTF/Maplestory OTF Light.otf", 20)
    except:
        content_font = ImageFont.load_default()

    draw.text((30, 130), f"[{nick}@{server}]", font=nick_font, fill=(0, 0, 0))

    draw.text((10, 185), job, font=content_font, fill=(0, 0, 0))

    draw.text((10, 210), f"레벨: {lv}({per}%)", font=content_font, fill=(0, 0, 0))

    draw.text((10, 235), f"길드: {guild} │ 인기도: {pop}", font=content_font, fill=(0, 0, 0))

    draw.text((10, 260), f"월드 랭킹: {comma(world_rank)}위", font=content_font, fill=(0, 0, 0))

    draw.text((10, 285), f"직업 랭킹(월드): {comma(class_rank)}위", font=content_font, fill=(0, 0, 0))

    # 6. 전송
    img_bytes = io.BytesIO()
    canvas.save(img_bytes, format="PNG")
    img_bytes.seek(0)
    return img_bytes

def maplegg(nick, sender):
    if nick is None or nick == "":
        return "닉네임을 입력해 주세요"
    else:
        recordnick(sender, nick)
        ocid = search_api_ocid(nick)
        search1 = search_maple_api(ocid)
        server = search1["world_name"]
        if "리부트" in server:
            search2 = search_maple_api_ranking(ocid, 1)
        else:
            search2 = search_maple_api_ranking(ocid, 0)

        per = search1["character_exp_rate"]
        img = search1["character_image"]

        job1 = search2["ranking"][0]["class_name"]
        job2 = search2["ranking"][0]["sub_class_name"]
        lv = int(search2["ranking"][0]["character_level"])
        pop = search2["ranking"][0]["character_popularity"]
        gu = search2["ranking"][0]["character_guildname"]
        ranking_all = int(search2["ranking"][0]["ranking"])

        history_db_save(nick, lv, "")
        
        if job2 is not None and job2 != "":
            printjob = f"{job1}/{job2}"
        else:
            printjob = f"{job1}/{job1}"

        if "리부트" in server:
            search3 = search_maple_api_ranking_class(ocid, job1, job2, 1)
        else:
            search3 = search_maple_api_ranking_class(ocid, job1, job2, 0)

        ranking_class = int(search3["ranking"][0]["ranking"])

        img_bytes = generate_character_image(img, nick, server, printjob, lv, per, gu, pop, ranking_all, ranking_class)

        text_print = "\n".join([
            "메애기/환산 링크",
            "\u200b" * 500,
            f"메애기 링크: https://meaegi.com/s/{parse.quote(nick)}",
            "",
            f"환산 링크: https://maplescouter.com/info?name={parse.quote(nick)}",
            "",
            "Special Thanks 정쿠, Lune"
        ])

        result_json = {
            "img_bytes":img_bytes,
            "text_print":text_print
        }

        return result_json
        

def handle_message(chat):
    if "@메이플" in chat.message.msg or "!메아플" in chat.message.msg:
        parts = chat.message.msg.split(" ")
        if len(parts) < 2:
            nick = recommendnick(chat.sender.name)
        else:
            nick = parts[1]
        try:
            res = maplegg(nick, chat.sender.name)
            chat.reply_media(res["img_bytes"])
            chat.reply(res["text_print"])
        except Exception as e:
            nick = chat.message.msg[5:]
            chat.reply(f"[{nick}]\n2023.12.21 이후 기록이 없는 캐릭터명 입니다.")
            raise
